File: tool/cli/internal/api/eval.go
package api

import (
	"context"
)

// EvalDataset / EvalCase / EvalRun / EvalResult 与后端 server/internal/api/eval.go 视图对齐。
type EvalDataset struct {
	ID            uint   `json:"id"`
	Name          string `json:"name"`
	Description   string `json:"description"`
	DefaultGrader string `json:"default_grader"`
	DefaultModel  string `json:"default_model"`
	CreatedAt     string `json:"created_at"`
}

type EvalCase struct {
	ID         uint   `json:"id"`
	DatasetID  uint   `json:"dataset_id"`
	Input      string `json:"input"`
	Expected   string `json:"expected"`
	Grader     string `json:"grader"`
	Model      string `json:"model"`
}

type EvalRun struct {
	ID            uint    `json:"id"`
	DatasetID     uint    `json:"dataset_id"`
	Model         string  `json:"model"`
	Grader        string  `json:"grader"`
	Repeats       int     `json:"repeats"`
	Status        string  `json:"status"`
	ScoreAvg      float64 `json:"score_avg"`
	PassRate      float64 `json:"pass_rate"`
	TotalCases    int     `json:"total_cases"`
	TotalAttempts int     `json:"total_attempts"`
	Error         string  `json:"error"`
	CreatedAt     string  `json:"created_at"`
}

type EvalResult struct {
	ID        uint    `json:"id"`
	RunID     uint    `json:"run_id"`
	CaseID    uint    `json:"case_id"`
	Attempt   int     `json:"attempt"`
	Grader    string  `json:"grader"`
	Output    string  `json:"output"`
	Score     float64 `json:"score"`
	Passed    bool    `json:"passed"`
	LatencyMs int64   `json:"latency_ms"`
	Error     string  `json:"error"`
}

// ListEvalDatasets 调 GET /api/eval/datasets。
func (c *Client) ListEvalDatasets(ctx context.Context) ([]EvalDataset, error) {
	var r struct {
		Datasets []EvalDataset `json:"datasets"`
	}
	if err := c.do(ctx, "GET", "/eval/datasets", nil, &r); err != nil {
		return nil, err
	}
	return r.Datasets, nil
}

// CreateEvalDataset 调 POST /api/eval/datasets。
func (c *Client) CreateEvalDataset(ctx context.Context, name, description, grader, model string) (*EvalDataset, error) {
	var r EvalDataset
	err := c.do(ctx, "POST", "/eval/datasets", map[string]string{
		"name":           name,
		"description":    description,
		"default_grader": grader,
		"default_model":  model,
	}, &r)
	return &r, err
}

// ListEvalCases 调 GET /api/eval/datasets/:id/cases。
func (c *Client) ListEvalCases(ctx context.Context, datasetID uint) ([]EvalCase, error) {
	var r struct {
		Cases []EvalCase `json:"cases"`
	}
	if err := c.do(ctx, "GET", "/eval/datasets/"+itoa(datasetID)+"/cases", nil, &r); err != nil {
		return nil, err
	}
	return r.Cases, nil
}

// RunEval 调 POST /api/eval/datasets/:id/run（异步；返回 running 记录）。
func (c *Client) RunEval(ctx context.Context, datasetID uint, model, grader string, repeats int) (*EvalRun, error) {
	body := map[string]any{"repeats": repeats}
	if model != "" {
		body["model"] = model
	}
	if grader != "" {
		body["grader"] = grader
	}
	var r EvalRun
	err := c.do(ctx, "POST", "/eval/datasets/"+itoa(datasetID)+"/run", body, &r)
	return &r, err
}

// ListEvalRuns 调 GET /api/eval/runs?dataset_id=。
func (c *Client) ListEvalRuns(ctx context.Context, datasetID uint) ([]EvalRun, error) {
	path := "/eval/runs"
	if datasetID != 0 {
		path += "?dataset_id=" + itoa(datasetID)
	}
	var r struct {
		Runs []EvalRun `json:"runs"`
	}
	if err := c.do(ctx, "GET", path, nil, &r); err != nil {
		return nil, err
	}
	return r.Runs, nil
}

// GetEvalRun 调 GET /api/eval/runs/:id。
func (c *Client) GetEvalRun(ctx context.Context, runID uint) (*EvalRun, error) {
	var r EvalRun
	if err := c.do(ctx, "GET", "/eval/runs/"+itoa(runID), nil, &r); err != nil {
		return nil, err
	}
	return &r, nil
}

// ListEvalResults 调 GET /api/eval/runs/:id/results。
func (c *Client) ListEvalResults(ctx context.Context, runID uint) ([]EvalResult, error) {
	var r struct {
		Results []EvalResult `json:"results"`
	}
	if err := c.do(ctx, "GET", "/eval/runs/"+itoa(runID)+"/results", nil, &r); err != nil {
		return nil, err
	}
	return r.Results, nil
}

// itoa 是避免引入 strconv 的极小整数转字符串（CLI 包内复用）。
func itoa(u uint) string {
	if u == 0 {
		return "0"
	}
	var b [20]byte
	i := len(b)
	for u > 0 {
		i--
		b[i] = byte('0' + u%10)
		u /= 10
	}
	return string(b[i:])
}

File: web/src/api/eval.ts
// 评估回归（M5-05）API 封装：评估集 / 用例 / 运行 / 结果 四区。
// 后端契约见 server/internal/api/eval.go（路径前缀 /eval，client 自动拼 /api）。
import { request } from './client'

// ---- 类型 ----

export type GraderType = 'exact' | 'contains' | 'llm'
export type EvalRunStatus = 'running' | 'done' | 'failed'

export interface EvalDataset {
  id: number
  user_id: number
  name: string
  description: string
  default_grader: GraderType
  default_model: string
  created_at: string
  updated_at: string
}

export interface EvalCase {
  id: number
  dataset_id: number
  input: string
  expected: string
  grader: GraderType
  model: string
  created_at: string
  updated_at: string
}

export interface EvalRun {
  id: number
  dataset_id: number
  user_id: number
  model: string
  grader: GraderType
  repeats: number
  status: EvalRunStatus
  score_avg: number
  pass_rate: number
  total_cases: number
  total_attempts: number
  error: string
  started_at: string | null
  finished_at: string | null
  created_at: string
}

export interface EvalResult {
  id: number
  run_id: number
  dataset_id: number
  case_id: number
  attempt: number
  grader: GraderType
  output: string
  score: number
  passed: boolean
  latency_ms: number
  error: string
  created_at: string
}

// ---- 评估集 Dataset ----

export interface ListEvalDatasetsResp {
  datasets: EvalDataset[]
  total: number
}

export async function listEvalDatasets(): Promise<ListEvalDatasetsResp> {
  return request<ListEvalDatasetsResp>('/eval/datasets')
}

export async function createEvalDataset(body: {
  name: string
  description?: string
  default_grader: GraderType
  default_model?: string
}): Promise<EvalDataset> {
  return request<EvalDataset>('/eval/datasets', { method: 'POST', body })
}

export async function updateEvalDataset(
  id: number,
  body: {
    name?: string
    description?: string
    default_grader?: GraderType
    default_model?: string
  },
): Promise<EvalDataset> {
  return request<EvalDataset>(`/eval/datasets/${id}`, { method: 'PUT', body })
}

export async function deleteEvalDataset(id: number): Promise<void> {
  await request(`/eval/datasets/${id}`, { method: 'DELETE' })
}

// ---- 用例 Case ----

export interface ListEvalCasesResp {
  cases: EvalCase[]
  total: number
}

export async function listEvalCases(datasetId: number): Promise<ListEvalCasesResp> {
  return request<ListEvalCasesResp>(`/eval/datasets/${datasetId}/cases`)
}

export async function createEvalCase(
  datasetId: number,
  body: { input: string; expected: string; grader?: GraderType; model?: string },
): Promise<EvalCase> {
  return request<EvalCase>(`/eval/datasets/${datasetId}/cases`, { method: 'POST', body })
}

export async function updateEvalCase(
  datasetId: number,
  caseId: number,
  body: { input?: string; expected?: string; grader?: GraderType; model?: string },
): Promise<EvalCase> {
  return request<EvalCase>(`/eval/datasets/${datasetId}/cases/${caseId}`, {
    method: 'PUT',
    body,
  })
}

export async function deleteEvalCase(datasetId: number, caseId: number): Promise<void> {
  await request(`/eval/datasets/${datasetId}/cases/${caseId}`, { method: 'DELETE' })
}

// ---- 运行 Run ----

export interface ListEvalRunsResp {
  runs: EvalRun[]
  total: number
}

export async function listEvalRuns(datasetId?: number): Promise<ListEvalRunsResp> {
  const qs = datasetId != null ? `?dataset_id=${datasetId}` : ''
  return request<ListEvalRunsResp>(`/eval/runs${qs}`)
}

export async function getEvalRun(id: number): Promise<EvalRun> {
  return request<EvalRun>(`/eval/runs/${id}`)
}

export async function runEval(
  datasetId: number,
  body?: { model?: string; grader?: GraderType; repeats?: number },
): Promise<EvalRun> {
  return request<EvalRun>(`/eval/datasets/${datasetId}/run`, { method: 'POST', body: body ?? {} })
}

// ---- 结果 Result ----

export interface ListEvalResultsResp {
  results: EvalResult[]
  total: number
}

export async function listEvalResults(runId: number): Promise<ListEvalResultsResp> {
  return request<ListEvalResultsResp>(`/eval/runs/${runId}/results`)
}
